protocol/superchainerc20/superc20Redesign.md

```diff
@@ -0,0 +1,101 @@
+    - This is necessary to conserve the invariant of the same address corresponding to the same (trusted) implementation, which is key to cross-chain access control.
+    - The OptimismSuperchainERC20 will be a Beacon Proxy. The OptimismSuperchainERC20 implementation would now be similar to the OptimismMintableERC20 implementation (give mint and burn rights to the L2StandardBridge and include a remoteToken ), but with an initialize function to replace the constructor.
+    - It is possible to reuse the `OptimismMintableERC20Factory` and avoid introducing a new predeploy factory.
+
+**Code modifications:**
+
+- The `OptimismMintableERC20Factory` will be upgraded to deploy `OptimismSuperchainERC20` , with the corresponding `init_code` as the `OptimismMintableERC20`s and using CREATE3.
+- The `BeaconContract` and `convert` function will work the same way.
+    - The `OptimismSuperchainERC20Factory` will of course be modified to deploy the new simpler implementations, but will still record deployments in a mapping.
+
+The updated flow would look like follows:
+
+```mermaid
+sequenceDiagram
+  participant from
+  participant L2SBA as L2StandardBridge (Chain A)
+  participant SuperERC20_A as SuperchainERC20 (Chain A)
+  participant Messenger_A as L2ToL2CrossDomainMessenger (Chain A)
+  participant Inbox as CrossL2Inbox
+  participant Messenger_B as L2ToL2CrossDomainMessenger (Chain B)
+  participant L2SBB as L2StandardBridge (Chain B)
+  participant SuperERC20_B as SuperchainERC20 (Chain B)
+  participant Recipient as to
+
+  from->>L2SBA: sendERC20To(tokenAddr, to, amount, chainID)
+  L2SBA->>SuperERC20_A: burn(from, amount)
+  L2SBA->>Messenger_A: sendMessage(chainId, message)
+  L2SBA-->L2SBA: emit SentERC20(tokenAddr, from, to, amount, destination)
+  Inbox->>Messenger_B: relayMessage()
+  Messenger_B->>L2SBB: relayERC20(tokenAddr, from, to, amount)
+  L2SBB->>SuperERC20_B: mint(to, amount)
+  L2SBB-->L2SBB: emit RelayedERC20(from, to, amount, source)
+
+```
+
+It’s important to notice that, even though this functions get implemented in the bridge, tokens will still have agency to reimplement the bridging functions on their contract if they want so. The alternative design gives the optionality.
+
+## Comparison
+
+### Pros of using the alternative design
+
+- **Compatible with the current design:** adding the `sendERC20` and `relayERC20` functions to the bridge does not imply that these should no longer exist in the token. It only gives the option to token developers not to include them.
+- **Separation of functionalities:** bridging lives in the bridge, token functions live in the token. This fosters clarity but also comes with improved upgrade flows.
+    - If bridging functions are implemented on the ERC20, updates to the bridge might require updates to every ERC20. This might be easy to do for implementations that use a BeaconProxy structure, but not in other cases.
```

Comment:
> If bridging functions are implemented on the ERC20, updates to the bridge might require updates to every ERC20. This might be easy to do for implementations that use a BeaconProxy structure, but not in other cases.

What do you mean here? I thought we do use the beacon pattern so I'm unsure what this bullet means exactly

Reply:
The idea is hinting towards the philosophy of not enshrining a particular superchain ERC20 factory, since there is no one fits all token contract. Rather any token simply needs to give minting rights to the new bridge predeploy to be superchain erc20 compatible. We could also enshrine the same address constraint as well, not sure if we should or not, i think we need to do a little implementation work to find out the tradeoffs

Reply:
We will use the Beacon patter for `OptimismSuperchainERC20`, aka tokens that convert back and forth with legacy. Those tokens are gonna be generated by the `OptimismSuperchainERC20Factory` predeploy.

But then for generic `SuperchainERC20` we will not enforce any pattern nor create an implementation specific factory. We think the best path is to suggest using the `Create2Deployer` preinstall with their `init_code`, but they can do whatever they want. Devs might implement a Beacon pattern or not. Or even use a Beacon but point to a different implementation.

In that scenario, a potential upgrade to the `L2toL2CrossDomainMessenger` for example would imply that every single implementation would need to be upgraded in order to be compatible; or would give us very little flexibility in potential changes to the `L2toL2CrossDomainMessenger`.
The new design doesn't have that problem.

To me, this is the most important pro of the list.

Reply:
I see now, thank you both. So now my thinking is that we should not encourage devs to implement bridging functions on their own ERC20, because:
- It might break depending on changes we make to `L2toL2CrossDomainMessenger`, or it may at least restrict what changes we feel comfortable making
- It fragments UX, which might confuse users as to why some tokens have a different entry point
- Added risk that the ERC20 methods are not implemented correctly

Is there value to having devs implement bridging functionality on the token?

Reply:
The only real value that I can think of at this point is the user would interact directly with the token rather than interacting with the bridge

Reply:
I hope "we should not encourage" does not mean "we give no choice".